File: pkg/_template-pool-factory/math-implementations/python/math_check.py
"""TEMPLATE: the entry point pvt/math-check loads. See pvt/math-check/README.md."""
from constant_sum_pool import ConstantSum

# Must equal the `poolType` returned by math-check/pool.ts.
POOL_TYPE = "CONSTANT_SUM"


def create_pool(pool):
    """`pool` is a BasePoolState with the pool-specific fields attached as snake_case attributes (here: pool.rate)."""
    return ConstantSum(pool)


# For a pool with a hook, also define:
#   HOOK_TYPE = "MyHook"
#   def create_hook(pool): return MyHook(), MyHookState(...)

File: pkg/_template-pool-factory/math-implementations/rust/src/bin/math_check.rs
//! TEMPLATE: runs this implementation against bush-maths test data.
//! `cargo run --features math-check --bin math_check -- <test-data-file.json>` (pvt/math-check does this for you).

use bush_maths::PoolBase;
use constant_sum_pool_maths::ConstantSumPool;
use math_check_support::{no_hook, PoolJson};

/// Build the pool maths from the `pool` block of a test-data file. `PoolJson` gives typed access to its fields.
fn make_pool(pool: &PoolJson) -> Result<Box<dyn PoolBase>, String> {
    Ok(Box::new(ConstantSumPool::new(pool.u("rate")?)))
}

fn main() {
    // For a pool with a hook, pass a `make_hook` returning `Some(Hook { hook_type, hook, state })` instead of `no_hook`.
    math_check_support::run(make_pool, no_hook);
}
